droidlet/parallel.py

```diff
@@ -0,0 +1,117 @@
+import traceback
+import queue
+from typing import Callable, List
+import cloudpickle
+# you're wondering wtf? why is numpy needed in this file?
+# it's a workaround for https://github.com/pytorch/pytorch/issues/37377
+import numpy 
+from torch import multiprocessing as mp
+from threading import Thread
+
+multiprocessing = mp.get_context("spawn")
+
+class Process(multiprocessing.Process):
+    """
+    Class which returns child Exceptions to Parent.
+    https://stackoverflow.com/a/33599967/4992248
+    """
+
+    def __init__(self, *args, **kwargs):
+        multiprocessing.Process.__init__(self, *args, **kwargs)
+        self._parent_conn, self._child_conn = multiprocessing.Pipe()
+        self._exception = None
+
+    def run(self):
+        try:
+            multiprocessing.Process.run(self)
+            self._child_conn.send(None)
+        except Exception as e:
+            tb = traceback.format_exc()
+            self._child_conn.send((e, tb))
+            raise e
+
+    @property
+    def exception(self):
+        if self._parent_conn.poll():
+            self._exception = self._parent_conn.recv()
+        return self._exception
+
+def _runner(_init_fn, init_args, _process_fn, shutdown_event, input_queue, output_queue):
```

Comment:
why is `_runner` defined here instead of being a private fn of `BackgroundTask`?

Reply:
because it needs to be picklable. A class-member is not picklable, but a global function is.

Reply:
on that note, for `init_fn` and `process_fn`, I am using `cloudpickle` precisely because I want to be able to pass class-local or function-local functions into `BackgroundTask` and they are only picklable either via `cloudpickle` or via `dill`, and not with the out of box pickler.